src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,12 +225,118 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+            // to still provide those items later before they are truly required.            
+
+            if (!options.ShouldShowItemsFromUnimportNamspaces())
+            {
+                // No need to trigger expanded providers at all if the feature is disabled, just trigger core providers and return;
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+
+            if (!session.TextView.Options.GetOptionValue(DefaultOptions.ResponsiveCompletionOptionId))
+            {
+                // We tie the behavior of delaying expand items to editor's "responsive completion" option.
+                // i.e. "responsive completion" disabled == always wait for all items to be calculated.
+                options = options with
+                {
+                    BlockOnExpandedCompletion = true,
+                    ExpandedCompletionBehavior = ExpandedCompletionMode.AllItems,
+                };
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation, options, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+
+            var expandedItemsTask = Task.Run(async () => await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly }, cancellationToken).ConfigureAwait(false),
+                cancellationToken);
+
+            // Record how long it takes for the background task to complete.
+            var stopwatch = Stopwatch.StartNew();
+            expandedItemsTask = expandedItemsTask.ContinueWith(task =>
+            {
+                AsyncCompletionLogger.LogAdditionalTicksToCompleteDelayedImportCompletionDataPoint((int)stopwatch.ElapsedMilliseconds);
+                return task.Result;
+            }, cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+
+            // Now trigger and wait for core providers to return;
+            var (nonExpandedContext, nonExpandedCompletionList) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+            AddPropertiesToSession(session, nonExpandedCompletionList, triggerLocation);
+
+            if (!expandedItemsTask.IsCompleted && !options.BlockOnExpandedCompletion)
```

Comment:
i'm confused by the option checking.  Isn't BlockOnExpandedCompletion known to be false since we didn't go into the !ResponsiveCompletionOptionId above?

Reply:
You are right, actually `BlockOnExpandedCompletion` is longer needed since we are using editor's option directly.